api/api.go

```diff
@@ -0,0 +1,321 @@
+        HandlerFunc(a.PostArticleSubscriptionStatus).Methods(http.MethodPost)
+
+    router.Path("/subscriptions/searches").Queries("query", "{query}").
+        HandlerFunc(a.GetSearchQuerySubscriptionStatus).Methods(http.MethodGet)
+    router.Path("/subscriptions/searches").Queries("query", "{query}", "subscribe", "{subscribe:(?:true|false)}").
+        HandlerFunc(a.PostSearchQuerySubscriptionStatus).Methods(http.MethodPost)
+
+    return router
+}
+
+func RespondWithJSON(w http.ResponseWriter, object interface{}, status int) error {
+    w.Header().Set("Content-Type", "application/json")
+
+    if object != nil { // non-empty body
+        if err := json.NewEncoder(w).Encode(object); err != nil {
+            w.WriteHeader(http.StatusInternalServerError)
+            return err
+        }
+    }
+
+    w.WriteHeader(status)
+    return nil
+}
+
+// postRegister handles request for a new account creation.
+func (a *UnarXivApi) PostRegister(w http.ResponseWriter, r *http.Request) {
+    var registerRequest usecases.AuthenticationRequest
+    if err := json.NewDecoder(r.Body).Decode(&registerRequest); err != nil {
+        w.WriteHeader(http.StatusBadRequest)
+        return
+    }
+
+    authData, err := a.Core.Register(&registerRequest)
+    if err != nil { // todo: map domain errors to http error codes
+        w.WriteHeader(http.StatusInternalServerError)
+        log.Printf("Error happened in Core.Register: %v", err)
+        return
+    }
+
+    if err := RespondWithJSON(w, &authData, http.StatusCreated); err != nil {
+        log.Printf("Error happened while responding to PostRegister: %v", err)
+    }
+}
+
+// PostLogin handles login request for existing user.
+func (a *UnarXivApi) PostLogin(w http.ResponseWriter, r *http.Request) {
+    var authRequest usecases.AuthenticationRequest
+    if err := json.NewDecoder(r.Body).Decode(&authRequest); err != nil {
+        w.WriteHeader(http.StatusBadRequest)
+        return
+    }
+
+    authData, err := a.Core.Login(&authRequest)
+    if err != nil { // todo: map domain errors to http error codes
+        w.WriteHeader(http.StatusInternalServerError)
+        log.Printf("Error happened in Core.Login: %v", err)
+        return
+    }
+
+    if err := RespondWithJSON(w, &authData, http.StatusOK); err != nil {
+        log.Printf("Error happened while responding to PostLogin: %v", err)
+    }
+}
+
+func (a *UnarXivApi) GetSearch(w http.ResponseWriter, r *http.Request) {
+    var searchQueryRequest = usecases.SearchQueryRequest{}
+    if err := r.ParseForm(); err != nil {
+        w.WriteHeader(http.StatusBadRequest)
+        log.Printf("Error happened while parsing form params: %v", err)
+        return
+    }
+    searchQueryRequest.Query = r.Form.Get("query")
+    offset, err := strconv.Atoi(r.Form.Get("offset"))
+    if err != nil && offset > 0 {
+        var u32offset uint32
+        u32offset = uint32(offset)
+        searchQueryRequest.Offset = &u32offset
```

Comment:
Why do you use pointer here? And why using additional variable?

Reply:
It is a pointer because it is actually an optional field. I guess in this particular case it should have a default zero value instead of being optional.

I didn't realize how to construct a reference to a uint32 and initialize it with int value in-place. Would something like `&uint32{uint32(offset)}` work?

Anyway, it is going to be rewritten.

Reply:
fixed